sourcegraph/sourcegraph
# Expose new Cody Subscriptions API & integrate rate limits with SSC subscription data

Follow-up PR of https://github.com/sourcegraph/sourcegraph/pull/59763

closes: https://github.com/sourcegraph/sourcegraph/issues/58123
closes: https://github.com/sourcegraph/sourcegraph/issues/59622
closes: https://github.com/sourcegraph/sourcegraph/issues/59620
closes: https://github.com/sourcegraph/sourcegraph/issues/59623

The previous PR:
- added a new SSC API client in dotcom to fetch user's subscription info
- added a func `cody.SubscriptionForUser` to be used to find cody subscription info for the user. 
- added unit tests

This PR:
- updates all the graphql resolvers wherever applicable to use the `cody.SubscriptionForUser` function rather than directly checking dotcom's DB for `users.cody_pro_enabled_at` column.
- updated the rate limits logic in `dotcom_user_resolver` to use cody.SubscriptionForUser` function as well. 
- Added the `currentUser.codySubscription` field to graphql api
```graphql
currentUser { 
  codySubscription { 
    status 
    plan 
    applyProRateLimtis 
    currentPeriodStartAt 
    currentPeriodEndAt 
  }
}
```
- Removed the following fields from the graphql API
```graphql
currentUser { 
  codyCurrentPeriodStartDate
  codyCurrentPeriodEndDate
}
```
Both the removed fields are not used by any of the Cody Clients and their usage is updated to follow the new schema for dotcom web. 

## Test plan
Unit tests present

**Manual testing plan**

1. Set up and run SAMS locally.
2. Integrate SAMS as an openidconnect auth provider with dotcom locally.
3. Create a user using the SAMS auth provider.
4. Enable feature flag use-ssc-for-cody-subscription from site admin.
5. Make a graphql API [request](https://sourcegraph.test:3443/api/console#%7B%22query%22%3A%22%7B%5Cn%20%20currentUser%20%7B%5Cn%20%20%20%20codyProEnabled%5Cn%20%20%7D%5Cn%7D%5Cn%22%7D) to confirm false.
6. Visit localhost:9992/cody and go through the subscription flow to sign up for Cody Pro.
7. Make a graphql API [request](https://sourcegraph.test:3443/api/console#%7B%22query%22%3A%22%7B%5Cn%20%20currentUser%20%7B%5Cn%20%20%20%20codyProEnabled%5Cn%20%20%7D%5Cn%7D%5Cn%22%7D) again to confirm true.
